### PR TITLE
Expose deactivation reason to user code and flow timeouts for activation/deactivation

### LINKED ISSUE #6384: Timeout ActivateAsync after a configurable duration
Grain activation time within the catalog is not currently bounded.
We ought to expose a configurable time for bounding activation times, cancelling the activation process and invalidating the grain activation if the time is exceeded.

Corresponding PR: #7261

### LINKED ISSUE #7324: Question: Getting the reason of a deactivation
When `OnDeactivateAsync` is called on a grain, is it possible to know why it got called? I would like to make a distinction whether it is caused by exceeding the idle time or if it is caused by a graceful shutdown of a silo. 

My goal is to reactivate a grain as soon as possible on a new silo after it got deactived due to a graceful shutdown of a silo. If, however, the actor got deactivated due to idle time it should *not* be activated automatically.

### PR DESCRIPTION
**Pull request by ReubenBond**
Replaces #7261
Fixes #6384
Fixes #7324

Exposes the reason for a grain being deactivated to the developer via an optional overload on the grain implementation.
It also adds an overload of `OnActivateAsync` which accepts a `CancellationToken`.

``` C#
public interface IGrainBase
{
    IGrainContext GrainContext { get; }
    Task OnActivateAsync(CancellationToken token) => default;
    Task OnDeactivateAsync(DeactivationReason reason, CancellationToken token) => default;
}
```

`DeactivationReason` contains 3 properties:
* `Exception`, which is usually null since most exceptions which would cause deactivation happen during activation but `OnDeactivateAsync` is only called if activation was successful. It is provided in the case of `InconsistentStateException`.
* `ReasonCode`, an enum with values `None`, `ShuttingDown`, `ActivationFailed`, `InternalFailure`, `ActivationIdle`, `ActivationUnresponsive`, `DuplicateActivation`, `ApplicationError`, `ApplicationRequested` (for explicit `DeactivateOnIdle` calls))
* `Description`, a textual description

The methods were added on a new interface, `IGrainBase` in preparation for POCO grains (#7360). `Grain` implements `IGrainBase`

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/7503)

### CODE REVIEW

File: src/Orleans.Core.Abstractions/Core/Grain.cs [hunk 1]
```diff
@@ -196,39 +202,48 @@ protected Task<List<IGrainReminder>> GetReminders()
...
-        }
+        /// <param name="cancellationToken">A cancellation token which signals when activation is being canceled.</param>
+        public virtual Task OnActivateAsync(CancellationToken cancellationToken) => OnActivateAsync();
+
+        /// <summary>
+        /// This method is called at the end of the process of activating a grain.
+        /// It is called before any messages have been dispatched to the grain.
+        /// For grains with declared persistent state, this method is called after the State property has been populated.
+        /// </summary>
+        public virtual Task OnActivateAsync() => Task.CompletedTask;
```
Comment (benjaminpetit) [inline src/Orleans.Core.Abstractions/Core/Grain.cs:234]: I am afraid that having two different overridable `OnActivateAsync` method will be confusing for people. (and it's the same for `OnDeactivateAsync`)

File: src/Orleans.Core.Abstractions/Core/Grain.cs [hunk 2]
```diff
@@ -72,7 +78,7 @@ private static void ThrowInvalidContext()
         /// This constructor is particularly useful for unit testing where test code can create a Grain and replace
         /// the IGrainIdentity and IGrainRuntime with test doubles (mocks/stubs).
         /// </summary>
-        protected Grain(IGrainRuntime runtime) : this()
+        protected Grain(IGrainContext grainContext, IGrainRuntime runtime) : this()
```
Comment (benjaminpetit) [inline src/Orleans.Core.Abstractions/Core/Grain.cs:81]: Shouldn't we use the `grainContext` passed in the constructor?

Comment (ReubenBond) [reply src/Orleans.Core.Abstractions/Core/Grain.cs:81]: Yes, you're right. I will flip the ctors around

File: src/Orleans.Core.Abstractions/Core/Grain.cs (hunk 1 above)
Comment (ReubenBond) [reply src/Orleans.Core.Abstractions/Core/Grain.cs:234]: I can delete these no-arg overloads

**Pull request merged**